File: pages/20260824170700_holding_topology_passing/analysis/ruby/passing_schedule.rb
require_relative "passing_notation"
require_relative "passing_route"

module PassingAnalysis
  Event = Struct.new(
    :beat, :height, :ball, :from_body, :from_contact, :from_hand,
    :to_body, :to_contact, :to_hand, :hold, :pass, :socket_index, :kind,
    keyword_init: true
  )
  Schedule = Struct.new(
    :pattern, :ball_count, :highest, :cycle_tosses, :cycle_length, :period, :hand_count, :body_count,
    :hand_period,
    keyword_init: true
  )

  module Scheduler
    module_function

    def schedule(source, hold_twos: true)
      pattern = Notation.parse(source)
      ball_count = Notation.object_count(pattern)
      raise ArgumentError, "passing object count must be a positive integer" unless
        ball_count.is_a?(Integer) && ball_count.positive?

      highest = pattern.throws.flatten.map(&:height).max
      period = pattern.throws.first.length
      hand_count = pattern.body_count * 2
      depth = [highest, 1].max
      landing = Array.new(hand_count) { Array.new(depth) { [] } }
      intro = (0...ball_count).to_a
      cycle_tosses = []
      init_complete = false
      beat = 0
      throw_contacts = pattern.starting_hands.dup
      start_key = nil
      cycle_length = 0

      2000.times do
        available = landing.map(&:shift)
        landing.each { |queue| queue << [] }
        recorded = []
        pattern.body_count.times do |body|
          contact = throw_contacts.fetch(body)
          tokens = pattern.throws.fetch(body).fetch(beat % period)
          recorded.concat(
            toss_from_contact(tokens, landing, available, intro, body, contact, pattern.body_count, beat, hold_twos, init_complete)
          )
        end
        raise ArgumentError, "prop landing with no toss at beat #{beat}" if available.any? { |queue| !queue.empty? }

        cycle_tosses.concat(recorded)
        if init_complete
          if start_key.nil?
            start_key = Marshal.dump(landing)
          elsif beat.positive? && (beat % period).zero? && Marshal.dump(landing) == start_key
            cycle_tosses.pop while !cycle_tosses.empty? && cycle_tosses.last.beat == beat
            cycle_length = beat
            break
          end
        elsif intro.empty? && ((beat + 1) % period).zero?
          init_complete = true
          beat = -1
        end
        beat += 1
        pattern.body_count.times { |body| throw_contacts[body] = 1 - throw_contacts[body] }
      end
      raise ArgumentError, "pattern did not repeat" if cycle_length.zero?

      rotated = rotate_to_starting_hands(cycle_tosses, cycle_length, pattern.starting_hands)
      Schedule.new(
        pattern: pattern,
        ball_count: ball_count,
        highest: highest,
        cycle_tosses: rotated,
        cycle_length: cycle_length,
        period: period,
        hand_count: hand_count,
        body_count: pattern.body_count,
        hand_period: 2
      )
    end

    def toss_from_contact(tokens, landing, available, intro, body, contact, body_count, beat, hold_twos, record)
      events = []
      from_hand = Route.global_contact(body, contact)
      tokens.each_with_index do |token, socket_index|
        if token.height.zero?
          raise ArgumentError, "prop landing on 0 toss at beat #{beat}" unless available.fetch(from_hand).empty?
          if record
            events << Event.new(
              beat: beat, height: 0, ball: nil, from_body: body, from_contact: contact, from_hand: from_hand,
              to_body: body, to_contact: contact, to_hand: from_hand, hold: false, pass: false,
              socket_index: socket_index, kind: "empty"
            )
          end
          next
        end
        ball = available.fetch(from_hand).shift || intro.shift
        raise ArgumentError, "no prop available at beat #{beat}" if ball.nil?

        to_body = Route.destination_body(token, body, body_count)
        to_contact = Route.destination_contact(token, contact)
        to_hand = Route.global_contact(to_body, to_contact)
        hold = hold_twos && token.height == 2 && from_hand == to_hand
        if record
          events << Event.new(
            beat: beat, height: token.height, ball: ball, from_body: body, from_contact: contact, from_hand: from_hand,
            to_body: to_body, to_contact: to_contact, to_hand: to_hand, hold: hold, pass: to_body != body,
            socket_index: socket_index, kind: hold ? "hold" : "throw"
          )
        end
        landing.fetch(to_hand).fetch(token.height - 1) << ball
      end
      events
    end

    def rotate_to_starting_hands(cycle_tosses, cycle_length, starting_hands)
      offset = (0...cycle_length).find do |candidate|
        packet = cycle_tosses.select { |event| event.beat == candidate }
        starting_hands.each_with_index.all? do |contact, body|
          event = packet.find { |toss| toss.from_body == body }
          event && event.from_contact == contact
        end
      end
      return cycle_tosses if offset.nil? || offset.zero?

      cycle_tosses.map do |event|
        event.dup.tap { |copy| copy.beat = (event.beat - offset + cycle_length) % cycle_length }
      end.sort_by { |event| [event.beat, event.from_hand] }
    end
  end
end

File: pages/20260824170700_holding_topology_passing/analysis/julia/passing_schedule.jl
function landing_key(landing)
    return [[copy(slot) for slot in queue] for queue in landing]
end

function rotate_to_starting_hands(cycle_tosses, cycle_length, starting_hands)
    offset = 0
    found = false
    for candidate in 0:(cycle_length - 1)
        packet = filter(event -> event.beat == candidate, cycle_tosses)
        aligned = all(enumerate(starting_hands)) do (body_index, contact)
            body = body_index - 1
            event = findfirst(toss -> toss.from_body == body, packet)
            event !== nothing && packet[event].from_contact == contact
        end
        if aligned
            offset = candidate
            found = true
            break
        end
    end
    found || return cycle_tosses
    rotated = [PassingEvent(
        mod(event.beat - offset, cycle_length), event.height, event.ball, event.from_body,
        event.from_contact, event.from_hand, event.to_body, event.to_contact, event.to_hand,
        event.hold, event.pass, event.socket_index, event.kind,
    ) for event in cycle_tosses]
    return sort(rotated, by=event -> (event.beat, event.from_hand))
end

function toss_from_contact!(
    tokens, landing, available, intro, body, contact, body_count, beat, hold_twos, record,
)
    events = PassingEvent[]
    from_hand = global_contact(body, contact)
    for (socket_index, token) in enumerate(tokens)
        if token.height == 0
            isempty(available[from_hand + 1]) ||
                throw(ArgumentError("prop landing on 0 toss at beat $beat"))
            record && push!(events, PassingEvent(
                beat, 0, nothing, body, contact, from_hand, body, contact, from_hand,
                false, false, socket_index - 1, "empty",
            ))
            continue
        end
        ball = isempty(available[from_hand + 1]) ? (isempty(intro) ? nothing : popfirst!(intro)) :
            popfirst!(available[from_hand + 1])
        ball === nothing && throw(ArgumentError("no prop available at beat $beat"))
        to_body = destination_body(token, body, body_count)
        to_contact = destination_contact(token, contact)
        to_hand = global_contact(to_body, to_contact)
        hold = hold_twos && token.height == 2 && from_hand == to_hand
        record && push!(events, PassingEvent(
            beat, token.height, ball, body, contact, from_hand, to_body, to_contact, to_hand,
            hold, to_body != body, socket_index - 1, hold ? "hold" : "throw",
        ))
        push!(landing[to_hand + 1][token.height], ball)
    end
    return events
end

function schedule_passing(source; hold_twos=true)
    pattern = parse_passing_siteswap(source)
    ball_count = passing_object_count(pattern)
    (ball_count isa Integer && ball_count >= 1) ||
        throw(ArgumentError("passing object count must be a positive integer"))
    highest = maximum(token.height for sequence in pattern.throws for multiplex in sequence for token in multiplex)
    period = length(pattern.throws[1])
    hand_count = pattern.body_count * 2
    depth = max(highest, 1)
    landing = [[Int[] for _ in 1:depth] for _ in 1:hand_count]
    intro = collect(0:(ball_count - 1))
    cycle_tosses = PassingEvent[]
    init_complete = false
    beat = 0
    throw_contacts = copy(pattern.starting_hands)
    start_key = nothing
    cycle_length = 0
    for _ in 1:2000
        available = [isempty(queue) ? Int[] : popfirst!(queue) for queue in landing]
        for queue in landing
            push!(queue, Int[])
        end
        recorded = PassingEvent[]
        for body in 0:(pattern.body_count - 1)
            contact = throw_contacts[body + 1]
            tokens = pattern.throws[body + 1][(beat % period) + 1]
            append!(recorded, toss_from_contact!(
                tokens, landing, available, intro, body, contact, pattern.body_count,
                beat, hold_twos, init_complete,
            ))
        end
        any(!isempty, available) && throw(ArgumentError("prop landing with no toss at beat $beat"))
        append!(cycle_tosses, recorded)
        if init_complete
            if start_key === nothing
                start_key = landing_key(landing)
            elseif beat > 0 && beat % period == 0 && landing_key(landing) == start_key
                while !isempty(cycle_tosses) && last(cycle_tosses).beat == beat
                    pop!(cycle_tosses)
                end
                cycle_length = beat
                break
            end
        elseif isempty(intro) && (beat + 1) % period == 0
            init_complete = true
            beat = -1
        end
        beat += 1
        for body in 1:pattern.body_count
            throw_contacts[body] = 1 - throw_contacts[body]
        end
    end
    cycle_length == 0 && throw(ArgumentError("pattern did not repeat"))
    rotated = rotate_to_starting_hands(cycle_tosses, cycle_length, pattern.starting_hands)
    return PassingSchedule(
        pattern, ball_count, highest, rotated, cycle_length, period, hand_count, pattern.body_count, 2,
    )
end
